### PR TITLE
Implement renew and pay

### LINKED ISSUE #12: As a librarian, I can extend a book’s loan
so that I can help borrowers to borrow the book for a longer period

### LINKED ISSUE #13: As a lazy librarian, I want to generate and record the fine of overdue books
so that I can keep track of overdue fines incurred by borrowers

### LINKED ISSUE #14: As a dutiful librarian, I want to record that a fine is paid
so that I can keep track of accounting and prevent duplicate payments

### LINKED ISSUE #55: Implement Pay command and functionality

### LINKED ISSUE #56: Implement Renew command and functionality

### PR DESCRIPTION
**Pull request by seeziyang**
Closes #55 
Closes #56 
Closes #12 
Closes #13 
Closes #14

### CODE REVIEW

**Comment by seeziyang** [comment]
Decided to not make return and renew support multiple indexes which complicates error response and changing data in model if users inputs a mix of valid and invalid indexes.

Will implement -all tag in another PR to return/renew all books at once.

**Comment by seeziyang** [comment]
Will need to integrate maximum renew count into user settings and set too.

File: src/main/java/seedu/address/commons/util/DateUtil.java [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
 
 import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
 import java.time.temporal.ChronoUnit;
 
 /**
  * Helper functions for handling LocalDates.
  */
 public class DateUtil {
+    private static DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
```
Comment (hoholyin) [inline src/main/java/seedu/address/commons/util/DateUtil.java:14]: Good use of pattern!

File: src/main/java/seedu/address/commons/util/StringUtil.java [hunk 2]
```diff
@@ -65,4 +65,30 @@ public static boolean isNonZeroUnsignedInteger(String s) {
...
+        requireNonNull(s);
+
+        String[] dollarsAndCents = s.split("\\."); // split by the dot '.'
+        if (dollarsAndCents.length > 1 && dollarsAndCents[1].length() > 2) {
+            return false;
+        }
+
+        try {
+            double value = Double.parseDouble(s);
+            if (value * 100.0 > (double) Integer.MAX_VALUE) { // value in cents is greater than max int value
```
Comment (hoholyin) [inline src/main/java/seedu/address/commons/util/StringUtil.java:86]: Maybe we can use the method `dollarToCents` from `FineUtil`

File: src/main/java/seedu/address/logic/commands/RenewCommand.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        if (!model.isServeMode()) {
+            throw new CommandException(MESSAGE_NOT_IN_SERVE_MODE);
+        }
+
+        List<Book> lastShownList = model.getFilteredBookList();
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(MESSAGE_INVALID_BOOK_DISPLAYED_INDEX);
+        }
+
+        Book bookToBeRenewed = lastShownList.get(index.getZeroBased()); // TODO change to second list index
```
Comment (hoholyin) [inline src/main/java/seedu/address/logic/commands/RenewCommand.java:68]: Is this TODO for this PR or future PRs?

File: src/main/java/seedu/address/logic/commands/RenewCommand.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        Loan loanToBeRenewed = bookToBeRenewed.getLoan().get();
+        Borrower servingBorrower = model.getServingBorrower();
+        if (!servingBorrower.hasCurrentLoan(loanToBeRenewed)) {
+            throw new CommandException(String.format(MESSAGE_NOT_LOANED_BY_BORROWER, servingBorrower, bookToBeRenewed));
+        }
+
+        // TODO take from userSettings in model instead
+        // TODO test this
```
Comment (hoholyin) [inline src/main/java/seedu/address/logic/commands/RenewCommand.java:80]: Likewise these 2 as well

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 5]
```diff
@@ -27,6 +27,8 @@
 public class ParserUtil {
 
     public static final String MESSAGE_INVALID_INDEX = "Index is not a non-zero unsigned integer.";
+    public static final String MESSAGE_INVALID_DOLLAR_AMOUNT =
+            "Dollar amount is not a non-zero number with at most 2 decimal places.";
```
Comment (hoholyin) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: Might be quite confusing to read. Maybe rephrase to something like:
Invalid dollar amount. Dollar amount should be a non-zero number with at most 2 decimal places.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 6]
```diff
@@ -154,6 +154,73 @@ public void addLoan(Loan loan) {
...
+
+        return change;
+    }
+
+    /**
+     * Private helper method to help iterate through each Loan in the returnedLoanList and then to update each Loan
+     * object with the maximum possible amount paid.
+     * Returns the leftover amount, i.e., the change.
+     */
+    private int payLoansFine(LoanList origReturnedLoanList, LoanList updatedReturnedLoanList, int amountInCents) {
```
Comment (hoholyin) [inline src/main/java/seedu/address/model/ModelManager.java:196]: Feels like a huge chunk of code with nested branches here, not sure if we can apply SLAP to make it easier to read

File: src/main/java/seedu/address/model/loan/Loan.java [hunk 7]
```diff
@@ -19,17 +19,31 @@
     private final BorrowerId borrowerId;
     private final LocalDate startDate;
     private final LocalDate dueDate;
-    // returnDate
-    // isReturned method
+    private final LocalDate returnDate;
+    private final int renewCount;
+    private final int remainingFineAmount; // in cents
+    private final int paidFineAmount; // in cents
```
Comment (hoholyin) [inline src/main/java/seedu/address/model/loan/Loan.java:25]: Would you consider naming the variables `remainingFineAmountInCents` and `paidFineAmountInCents`so we dont need the comment to clarify? Might be abit long for a name but I feel it might be easier when we are handling this variable next time too!

File: src/test/java/seedu/address/commons/util/DateUtilTest.java [hunk 8]
```diff
@@ -81,4 +82,31 @@ public void getNumOfDaysBetween_invalidDates_throwsNullPointerException() {
...
+                LocalDate.of(2019, 10, 12)), 0);
+    }
+
+    @Test
+    public void getNumOfDaysOverdue_invalidDate_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> DateUtil.getNumOfDaysOverdue(null, LocalDate.now()));
+    }
+
+    @Test
+    public void isDateBeforeToday() {
```
Comment (hoholyin) [inline src/test/java/seedu/address/commons/util/DateUtilTest.java:100]: Good tests coverage on boundary values!

File: src/test/java/seedu/address/logic/parser/CatalogParserTest.java [hunk 9]
```diff
@@ -65,7 +70,7 @@ public void parseCommand_add() throws Exception {
     @Test
     public void parseCommand_clear() throws Exception {
         assertTrue(parser.parseCommand(ClearCommand.COMMAND_WORD) instanceof ClearCommand);
-        assertTrue(parser.parseCommand(ClearCommand.COMMAND_WORD + " 3") instanceof ClearCommand);
+        assertTrue(parser.parseCommand(ClearCommand.COMMAND_WORD + SPACE_AND_VALID_INDEX) instanceof ClearCommand);
```
Comment (hoholyin) [inline src/test/java/seedu/address/logic/parser/CatalogParserTest.java:73]: Beautiful abstraction!

**Review by hoholyin — COMMENTED** [review]
Just a few comments here and there but generally looks great!

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 5 above)
Comment (seeziyang) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: Oops, it's a typo, my bad

File: src/main/java/seedu/address/commons/util/StringUtil.java (hunk 2 above)
Comment (seeziyang) [reply src/main/java/seedu/address/commons/util/StringUtil.java:86]: That method returns an int but the value here can be larger than max int value which is what I'm testing

File: src/main/java/seedu/address/logic/commands/RenewCommand.java (hunk 3 above)
Comment (seeziyang) [reply src/main/java/seedu/address/logic/commands/RenewCommand.java:68]: Future PRs!

Comment (seeziyang) [reply src/main/java/seedu/address/logic/commands/RenewCommand.java:68]: Will have to discuss with @Cronyxx too

File: src/main/java/seedu/address/model/loan/Loan.java (hunk 7 above)
Comment (seeziyang) [reply src/main/java/seedu/address/model/loan/Loan.java:25]: Actually I think I will keep this as is for now since it's an int

**Review by hoholyin — APPROVED** [review]
👍
